larpandoracontent/LArHelpers/LArHierarchyHelper.h

```diff
@@ -0,0 +1,629 @@
+         *
+         *  @return The vector of sub-threshold matches
+         */
+        const MCMatchesVector &GetSubThresholdMatches() const { return m_subThresholdMatches; }
+
+        /**
+         *  @brief  Retrieve the vector of unmatched MC nodes
+         *
+         *  @return The vector of unmatched MC
+         */
+        const MCHierarchy::NodeVector &GetUnmatchedMC() const { return m_unmatchedMC; };
+
+        /**
+         *  @brief  Retrieve the vector of unmatched reco nodes
+         */
+        const RecoHierarchy::NodeVector &GetUnmatchedReco() const { return m_unmatchedReco; }
+
+    private:
+        MCMatchesVector m_goodMatches;  ///< The vector of good matches
+        MCMatchesVector m_subThresholdMatches;  ///< The vector of matches MC nodes that don't pass quality cuts
+        MCHierarchy::NodeVector m_unmatchedMC;  ///< The vector of unmatched MC nodes
+        RecoHierarchy::NodeVector m_unmatchedReco;  ///< The vector of unmatched reco nodes
+        QualityCuts m_qualityCuts;  ///< The quality cuts to be applied to matches
+    };
+
+    /**
+     *  @brief  Fill an MC hierarchy based on the specified folding criteria (see MCHierarchy::Construct for details)
+     *
+     *  @param  mcParticleList The MCParticle list to use to fill this hierarchy
+     *  @param  caloHitList The list of CaloHits to use to fill this hierarchy
+     *  @param  foldToPrimaries Whether or not to fold to primary particles
+     *  @param  foldToLeadingShowers Whether or not to fold to leading shower particles
+     *  @param  hierarchy The output MC hierarchy
+     */
+    static void FillMCHierarchy(const pandora::MCParticleList &mcParticleList, const pandora::CaloHitList &caloHitList,
+        const bool foldToPrimaries, const bool foldToLeadingShowers, MCHierarchy &hierarchy);
+
+    /**
+     *  @brief  Fill a reconstructed hierarchy based on the specified folding criteria (see RecoHierarchy::Construct for details)
+     *
+     *  @param  pfoList The ParticleFlowObject list to use to fill this hierarchy
+     *  @param  foldToPrimaries Whether or not to fold to primary particles
+     *  @param  foldToLeadingShowers Whether or not to fold to leading shower particles
+     *  @param  hierarchy The output reconstructed hierarchy
+     */
+    static void FillRecoHierarchy(const pandora::PfoList &pfoList, const bool foldToPrimaries, const bool foldToLeadingShowers,
+        RecoHierarchy &hierarchy);
+
+    /**
+     *  @brief  Finds the matches between reconstructed and MC hierarchies.
+     *
+     *  @param  mcHierarchy The MC hiearchy
+     *  @param  recoHierarchy The reconstructed hierarchy
+     *  @param  matchInfo The output match information
+     */
+    static void MatchHierarchies(const MCHierarchy &mcHierarchy, const RecoHierarchy &recoHierarchy, MatchInfo &matchInfo);
+
+private:
+    typedef std::set<const pandora::MCParticle*> MCParticleSet;
+    typedef std::set<const pandora::ParticleFlowObject*> PfoSet;
```

Comment:
No, this is the first such definition (partly why they're private, in case we do want to define them alongside similar types to be more generally available)